### PR TITLE
Adding the PySpark script for Dataframe Transformation

### PR DESCRIPTION
**Pull request by ahmadnazeri**
# Description
This PR adds PySpark script to allow for dataframe transformations to be executed. 

### Select type(s) of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected)
          - This will break the Spark SQL provider because it introduces an additional parameter to the args. I will make the changes to this PR to avoid the breaking change.   
- [ ] Documentation update

# Checklist:

- [X] I have performed a self-review of my code
- [X] I have commented my code, particularly in hard-to-understand areas
- [X] I have made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] New and existing unit tests pass locally with my changes
- [x] I have fixed any merge conflicts

### CODE REVIEW

File: provider/scripts/offline_store_spark_runner.py [hunk 1]
```diff
@@ -1,6 +1,9 @@
+import types
 import argparse
+from typing import List
 from datetime import datetime
 
+# import dill
```
Comment (Sami1309) [inline provider/scripts/offline_store_spark_runner.py:6]: keep this comment?

File: provider/scripts/offline_store_spark_runner.py [hunk 2]
```diff
@@ -10,6 +13,13 @@
 # sql_query: eg. SELECT * FROM source_1 INNER JOIN source_2 ON source_1.id = source_2.id
```
Comment (Sami1309) [inline provider/scripts/offline_store_spark_runner.py:13]: update comment to reflect changes to arg parsing

File: provider/scripts/offline_store_spark_runner.py [hunk 3]
```diff
@@ -23,20 +33,67 @@ def execute_sql_query(job_type, output_uri, sql_query, source_list):
...
             output_dataframe.coalesce(1).write.option("header", "true").mode("overwrite").parquet(output_uri_with_timestamp)
-    except Exception as ident:
-        print(ident)
+            return output_uri_with_timestamp
+    except Exception as e:
+        print(e)
+        raise e
+
+def execute_df_job(output_uri, code, sources):
+    spark = SparkSession.builder.appName("Dataframe Transformation").getOrCreate()
```
Comment (Sami1309) [inline provider/scripts/offline_store_spark_runner.py:42]: use with to handle errors for the session?

File: provider/spark.go [hunk 4]
```diff
@@ -829,6 +836,31 @@ func updateQuery(query string, mapping []SourceMapping) (string, []string, error
 	return updatedQuery, sources, nil
 }
 
+func (spark *SparkOfflineStore) getSourcePath(path string) (string, error) {
+	filePaths := strings.Split(path[len("s3://"):], "/")
```
Comment (Sami1309) [inline provider/spark.go:840]: make this splitting and parsing its own function

**Review by Sami1309 — COMMENTED** [review]
mostly lgtm

File: provider/scripts/offline_store_spark_runner.py [hunk 5]
```diff
@@ -23,20 +33,67 @@ def execute_sql_query(job_type, output_uri, sql_query, source_list):
...
+    
+    func_parameters = {}
+    for name, location in sources.items():
+        func_parameters[name] = spark.read.parquet(location)
     
+    try:
+        # code = dill.loads(bytes(code))
+        func = types.FunctionType(code, globals(), "df_transformation")
```
Comment (sdreyer) [inline provider/scripts/offline_store_spark_runner.py:50]: What does this look like? Is it the entire transformation function or just the dataframe it returns?

File: provider/spark.go [hunk 6]
```diff
@@ -829,6 +836,31 @@ func updateQuery(query string, mapping []SourceMapping) (string, []string, error
...
+	fileType, fileName, fileVariant := filePaths[2], filePaths[3], filePaths[4]
```
Comment (sdreyer) [inline provider/spark.go:841]: Is this the format we store this? Or can a user give this path?

File: provider/scripts/offline_store_spark_runner.py (hunk 5 above)
Comment (ahmadnazeri) [reply provider/scripts/offline_store_spark_runner.py:50]: it is the entire transformation function. Here is the [test case](https://github.com/featureform/featureform/blob/feature/spark-df-transformations-pyscript/provider/scripts/tests/conftest.py#L71-L75).

File: provider/spark.go (hunk 6 above)
Comment (ahmadnazeri) [reply provider/spark.go:841]: It is a format we store.

File: provider/scripts/offline_store_spark_runner.py [hunk 7]
```diff
@@ -1,14 +1,17 @@
...
 from datetime import datetime
 
 from pyspark.sql import SparkSession
 
-# Expected format
 
-# job_type: "Transformation", "Materialization", or "Training Set",
-# output_uri: s3://featureform/{type}/{name}/{variant}
-# sql_query: eg. SELECT * FROM source_1 INNER JOIN source_2 ON source_1.id = source_2.id
-# source_list: [s3://featureform/{type}/{name}/{variant}, s3://featureform/{type}/{name}/{variant}...] (implicitly mapped to source_1, source_2, etc)
```
Comment (sdreyer) [inline provider/scripts/offline_store_spark_runner.py:11]: Kinda liked this as reference

**Review by sdreyer — APPROVED** [review]

**Pull request merged**